internal/auth/interface.go

```diff
@@ -0,0 +1,197 @@
+	// not be unique or stable over time, though they should be unique at any
+	// particular time.  For example, Alex may change their handle from
+	// "alexsmith" to "alexawesome" after a life change, and someone else might
+	// enroll the "alexsmith" handle.  If you are storing data, you want UserID,
+	// not HumanName.  If you are presenting data, you probably want HumanName.
+	//
+	// For KeyCloak, this is `preferred_username`.
+	HumanName string
+	// Provider is the identity provider that vended this identity.  Note that
+	// UserID and HumanName are only unique within the context of a single
+	// identity provider.
+	Provider IdentityProvider
+}
+
+// String implements strings.Stringer, and also provides a stable storage
+// representation of the Identity.
+func (i *Identity) String() string {
+	if i == nil {
+		return ""
+	}
+	p := i.Provider.String()
+	if p == "" {
+		// Special case for provider registered as "".
+		return i.UserID
+	}
+	return fmt.Sprintf("%s/%s", i.UserID, p)
+
+}
+
+// Human returns a human-readable representation of the identity, suitable for
+// presentation to humans.
+func (i *Identity) Human() string {
+	if i == nil {
+		return "<unknown>"
+	}
+	p := i.Provider.String()
+	if p == "" {
+		// Special case for provider registered as "".
+		return i.UserID
+	}
+	return fmt.Sprintf("%s/%s", i.HumanName, p)
+}
+
+// Resolver is an interface for resolving human-readable or stable identifiers
+// from either JWTs or stored strings
+type Resolver interface {
+
+	// Validate validates a token and returns an underlying identity representation
+	// suitable for use in authz calls.  This _probably_ reads data from the token,
+	// but could fetch from an external provider.
+	Validate(ctx context.Context, token jwt.Token) (*Identity, error)
+
+	// Resolve takes either a human-readable identifier or a stable identifier and
+	// returns the underlying identity.  This may involve looking up or defining
+	// the identity in the remote identity provider.
+	//
+	// For Keycloak + GitHub, this may define a new user in Keycloak based on
+	// GitHub user data if the user is not already known to Keycloak.
+	Resolve(ctx context.Context, id string) (*Identity, error)
+}
+
+// IdentityProvider provides an abstract interface for looking up identities
+// in a remote identity provider.
+type IdentityProvider interface {
+	Resolver
+
+	// String returns the name of the identity provider.  This should be a short
+	// one-word string suitable for presentation.  As a special case, a _single_
+	// provider may use the empty string as its name to act as a default / fallback
+	// provider.
+	String() string
+	// URL returns the `iss` URL of the identity provider.
+	URL() url.URL
+}
+
+// IdentityClient supports the ability to look up identities in one or more
+// IdentityProviders.
+type IdentityClient struct {
+	// This map is a bit overloaded; it maps both short provider names
+	providers map[string]IdentityProvider
```

Comment:
you could use `xsync.MapOf` which we already use in other parts of the code-base.